wallet/src/repository/wallet/mod.rs

```diff
@@ -648,116 +570,96 @@ where
-                .ok_or_else(|| Error::BalanceOverflow)?;
-            let txn_next_id = txn_id
-                .checked_add(1)
-                .ok_or_else(|| Error::TransactionValueOverflow)?;
-            let mut db_utxo_set: model::UtxoSet = self
-                .db
-                .get(&keys::account_utxo_set(account))
-                .unwrap_or_default();
-            let address = pkh.bech32(if self.params.testnet {
-                Environment::Testnet1
-            } else {
-                Environment::Mainnet
-            });
-            let key_balance = model::KeyBalance { pkh, amount };
-
-            match db_utxo_set.insert(out_ptr.clone(), key_balance.clone()) {
-                None => {
-                    log::info!(
-                        "Found transaction to our address {}! Amount: +{} satowits",
-                        address,
-                        amount
-                    );
-                }
-                Some(x) => {
-                    if x != key_balance {
+        let mut output_amount: u64 = 0;
+        for (index, output) in txn.outputs.iter().enumerate() {
+            if let Some(model::Path { .. }) = self.db.get_opt(&keys::pkh(&output.pkh))? {
+                let address = output.pkh.bech32(if self.params.testnet {
+                    Environment::Testnet1
+                } else {
+                    Environment::Mainnet
+                });
+                let key_balance = model::KeyBalance {
+                    pkh: output.pkh,
+                    amount: output.value,
+                };
+
+                let out_ptr = model::OutPtr {
+                    txn_hash: Vec::from(txn_hash),
+                    output_index: index as u32,
+                };
+
+                match db_utxo_set.insert(out_ptr.clone(), key_balance.clone()) {
+                    Some(x) if x == key_balance => {}
+                    _ => {
                         log::info!(
                             "Found transaction to our address {}! Amount: +{} satowits",
                             address,
-                            amount
+                            output.value
                         );
                     }
                 }
+                state.utxo_set.insert(out_ptr, key_balance);
+
+                output_amount = output_amount
+                    .checked_add(output.value)
+                    .ok_or_else(|| Error::TransactionBalanceOverflow)?;
             }
+        }
 
-            batch.put(&keys::transaction_value(account, txn_id), amount)?;
-            batch.put(
-                keys::transaction_entry(account, txn_id),
-                model::TransactionEntry::Credit,
-            )?;
+        let txn_balance = (output_amount as i128)
+            .checked_sub(input_amount as i128)
+            .ok_or_else(|| Error::TransactionBalanceUnderflow)?;
+
+        let txn_balance =
+            i64::try_from(txn_balance).map_err(|_| Error::TransactionValueOverflow)?;
```

Comment:
Shouldn't this be transaction balance?

```suggestion
            i64::try_from(txn_balance).map_err(|_| Error::TransactionBalanceOverflow)?;
```